### ISSUE TITLE
Faster binary data writing to socket

### ISSUE THREAD

**Issue by streamich** [uNetworking/uWebSockets.js#894]
uWebSockets.js accepts `ArrayBuffer`-and-co JavaScript binary formats when writing to the socket, that is great, but is not enough.

All data encoders pre-allocate more memory than needed, but current API does not allow to effectively provide only the slice of the final encoded data, instead of creating a new buffer-like object.

A simplified example:

```ts
const uint8 = new Uint8Array(1024);
const size = encodeMessage(uint8, message);
const output = uit8.subarray(0, size); // <--------- This is a problem, .subarray() is very slow.

res.end(output);
```

Instead, responses could be sped up substantially (and consume less memory), with an API as follows:

```ts
const uint8 = new Uint8Array(1024);
const size = encodeMessage(uint8, message);

res.end(uint8.buffer, 0, size);
```

, where `0` is offset and `size` is the length of binary chunk to copy. (No call to `.subarray()`, which is very slow.)

Constructing `ArrayBuffer`, or `Uint8Array`, or `Buffer` are very, very slow in Node.js and every release, at least since Node v14, it is getting slower.

I would like to propose an optimized API, where the binary chunk slice can be specified using `offset` and `length` params, instead of creating a temporary `ArrayBuffer` or `Uint8Array`.

Below are three options for this new API.

## Option 1

This would be the fastest option:

```ts
interface HttpResponse {
  endFast(buf: ArrayBuffer, offset: number, length: number): void;
  writeFast(buf: ArrayBuffer, offset: number, length: number): void;
}
```

The `.end()` and `.write()` are the most performance critical methods, so `.writeHeader()` and `.writeStatus()` could be ignored here.

##  Option 2

This option is slower, but would integrate nicely with the existing API:

```ts
type RecognizedString = RecognizedString | Slice;

interface Slice {
  buf: ArrayBuffer;
  offset: number;
  length: number;
}

// or

type Slice = [buf: ArrayBuffer, offset: number, length: number];
```

constructing a new instance of `Slice` is about 30x faster than creating a "slice" using `ArrayBuffer` or `Uint8Array`.

In TypeScript:

```ts
class Slice {
  constructor (public readonly buf: ArrayBuffer, public readonly offset: number, public readonly length: number) {}
}
const slice = new Slice(uint8.buffer, 0, size);

// or

type Slice = [buf: ArrayBuffer, offset: number, length: number];
const slice = [uint8.buffer, 0, size];
```

This option would integrate well also with the `.writeHeader` and `.writeStatus` methods.

`uWebSockets` could provide the `Slice` class, which V8 would hopefully compile as hidden class:

```ts
import {Slice} from 'uWebSockets.js';

const slice = new Slice(uint8.buffer, 0, size);
```

## Option 3

This option is similar to Option 2, but instead of extending the `RecognizedString` type, various existing methods would receive optional `offset` and `length` parameters:

```ts
interface HttpResponse {
  write(data: RecognizedString): number;
  write(data: RecognizedString, offset: number, length: number): number;

  end(body?: RecognizedString, closeConnection?: boolean, offset?: number, length?: number): HttpResponse;
}
```

---

The same applies to the WebSocket instances.

**Comment by streamich** [comment]
Here is a simple benchmark, which shows how slow creating a buffer slice using a bufferl-like objects is:

<img width="1277" alt="image" src="https://user-images.githubusercontent.com/9773803/235355093-1133c05f-55bb-44f3-b289-d10f4ebdf5f9.png">


Note:

- From Node v14 to Node v20, speed of instantiating a new `Uint8Array` is getting slower.
- Speed of creating a `ArrayBuffer` is in `s`, not `ms`
- Creating a new `Buffer` is not in the benchmark, but is about the same to a bit slower than `Uint8Array`.

**Comment by e3dio** [comment]
Please add to your benchmark the new Node v20 ArrayBuffer resize method https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer/resize
Pretty sure that is all you need

**Comment by streamich** [comment]
Thanks @e3dio, good to know about `ArrayBuffer.ptototype.resize()`, unfortunately it is the worst performing one:

<img width="1213" alt="image" src="https://user-images.githubusercontent.com/9773803/235356155-c524221c-3d00-44f2-94b2-9225fe5e47b8.png">

And it does not solve the problem, as:

- That method assumes that `offset` is always `0`, which is not the case.
- That method assumes that you own the whole `ArrayBuffer`, which also is not the case, the message encoder will share one large `ArrayBuffer`, encode multiple messages into it and hand out the slices.

**Comment by streamich** [comment]
Created a better benchmark for `.resize()`. In this benchmark the buffer is only resized down:

<img width="1229" alt="image" src="https://user-images.githubusercontent.com/9773803/235356967-8e734441-d87d-43c1-accc-c96a648331e3.png">

Still, `.resize()` is about as slow as creating a new `Uint8Array`. About an order of magnitude slower than creating a `Slice`.

**Comment by streamich** [comment]
`res.endFast(buf, offset, length)` would be even faster than the `new Slice()` method. Essentially if we only added these two methods, it would solve 90% of the problem:

```ts
interface HttpResponse {
  endFast(buf: ArrayBuffer, offset: number, length: number): void;
}

interface WebSocket {
  sendFast(buf: ArrayBuffer, offset: number, length: number, isBinary: boolean): void;
}
```

**Comment by uasan** [comment]
You want to use a normal buffer as a shared buffer, but how do you deal with race conditions when the uWS reads your buffer in a separate thread, are you already overwriting this buffer slice in the node process?

**Comment by e3dio** [comment]
Yes creating a resizable ArrayBuffer is slow and resize is also slow, that is disappointing. This is the fastest way I see, I don't see this in your benchmark:

```javascript
const buf = Buffer.allocUnsafe(1000); // initial fast buffer
const buf2 = buf.subarray(0,500); // zero copy buffer view of data
```

**Comment by uasan** [comment]
The subarray method returns a ref to a buffer slice, this is not a copy of the buffer, as I wrote above, without solving the race problems, you will not be able to use one buffer for many responses, I advise you to study the performance of the new method that returns a copy of the buffer

[Buffer.copyBytesFrom](https://nodejs.org/dist/latest-v20.x/docs/api/buffer.html#static-method-buffercopybytesfromview-offset-length)

**Comment by e3dio** [comment]
uWS send/end/write takes a Buffer, there is nothing mentioned about a Buffer view not working, I think uWS copies data on method call so `buf.subarray()` should work. However what @streamich is proposing would not work async for the reasons you mentioned

**Comment by streamich** [comment]
`.subarray()` is very slow. It is almost identical to `new Uint8Array()` and you can see performance of that in the screenshots above.

**Comment by streamich** [comment]
@uasan yes, my understanding is that uWS makes a copy of the passed in data.

**Comment by e3dio** [comment]
> .subarray() is very slow. It is almost identical to new Uint8Array()

I don't see subarray in your benchmark. My benchmark shows allocUnsafe+subarray is 3x faster than allocUnsafe+Uint8Array

**Comment by streamich** [comment]
@e3dio included `.subarray()`:

<img width="1196" alt="image" src="https://user-images.githubusercontent.com/9773803/235364529-b5a68d6b-328b-43e0-bf48-d2af8c373f22.png">

**Comment by streamich** [comment]
Creating a temporary `Uint8Array` just to represent a "slice" is a lot of unnecessary memory usage, [see in this StackOverflow answer](https://stackoverflow.com/questions/45803829/memory-overhead-of-typed-arrays-vs-strings) how much fields each `Uint8Array` instance holds.

**Comment by streamich** [comment]
Just to summarize my case:

1. In a non-trivial performance-cautious application there will be an `encoder`, which encodes responses into binary format.
2. No encoder will return you the exact `ArrayBuffer` of the message.
   - Typically, the message will be just some slice of a larger shared `Uint8Array`.
   - Or, if the encoder returns your own `ArrayBuffer` just for your message, it will be oversized. Because encoders initially allocate more space as they don't know the size of encoded data ahead of time.
3. So, currently, there is no efficient way to return the exact "slice".
   - Currently the best way is to use `.subarray()` which is very slow and allocates 208 bytes for __every message__.


```ts
uws.get('/rpc', (res, req) => {
  const result = rpc.exec(/* ... */);
  const slice = encoder.encode(result);

  // very slow:
  res.end(slice.buf.subarray(slice.start, slice.end));

  // would be nice to have:
  res.endFast(slice);
})
```

**Comment by e3dio** [comment]
@streamich your benchmark is not consistent because you use ArrayBuffer for new Uint8Array and Uint8Array for subarray. You need to start with either a Buffer or Uint8Array for each. When you start with either of those I show `buf.subarray()` is more than 10x faster than `new Uint8Array()`

**Comment by streamich** [comment]
@e3dio I'm not sure I understand what you mean. Here, `Buffer.prototype.subarray()` is even 2x slower than `Uint8Array.prototype.subarray()`:

<img width="1175" alt="image" src="https://user-images.githubusercontent.com/9773803/235366702-68b6b81e-880e-415e-949a-a47466dcf7a8.png">

**Comment by streamich** [comment]
Added `res.end()` case to show the performance impact of `res.end(new Slice())` over `res.end(buf, offset, length)`:

<img width="1297" alt="image" src="https://user-images.githubusercontent.com/9773803/235367011-49a288e7-6085-4138-9462-55e13b5af35d.png">

**Comment by uNetworkingAB** [comment]
I get your frustration with how slow and inefficient Node.js is implemented. Creating a slice of an ArrayBuffer is a core language feature and is comparable with std::string_view in C++, which is free of overhead. It should be close to free of cost, anywhere it is implemented.

The fact you can implement 20x faster slices on your own, as compared to what the Node.js core developers have done, is the tip of an iceberg summary of my past 7 years work with trying to get Node.js core developers to understand just how much they missed the mark, everywhere. But the only outcome my work resulted in was a permanent ban / block for being "a spreader of hate speech". I'm not woke enough to be able to convey a negative story, compatible with their safe spaces.

Of course, any sane implementation of a slice that immediately dies would use 0 memory and incur an insignificant cost as it would be caught by "most objects die young" GC strategy.

So now you have found that if uWS takes on the task of implementing a slicing of ArrayBuffer, by passing those extra arguments, you get a lot better performance. It doesn't surprise me at all, but uWS is a server library - it does not have the task of reimplementing core features of Node.js / JavaScript such as how ArrayBuffer is implemented. If JavaScript the language, and its very core features, cannot slice an array efficiently, then that is a problem for relevant party.

uWS could reimplement ArrayBuffer much more efficiently, but that's out of scope. We say that "you cannot do magic with your knees" and this is a perfect example. I cannot continue to lose my mind, reimplementing everything that is fundamentally broken in Node.js, that's why you have both Bun and Deno reimplement everything from scratch, since they have made the analysis that Node.js is beyond repair. It is broken to the core.

**Comment by e3dio** [comment]
> I'm not sure I understand what you mean

@streamich In your `new Uint8Array()` test you are starting from an ArrayBuffer and not a Buffer or Uint8Array which you used for the other tests. You need to start with the same Buffer or Uint8Array, which will slow it down. I show subarray() is 10x faster, you can confirm this

**Comment by ronag** [comment]
If you need a node `Buffer` there is a little trick you an use:

```js
const FastBuffer = Buffer[Symbol.species]

const buf2 = new FastBuffer(buf1.buffer, buf1.byteOffset + offset, length) // buf1.subarray(offset ,length)
```

Or something along those lines.

**Comment by ronag** [comment]
>  as compared to what the Node.js core developers have done

To be fair... this is a problem with `UInt8Array` which Node doesn't implement, it's V8.

**Comment by uasan** [comment]
Using the cork handler is important for performance, which means you need to take care of slice immutability before calling cork, i.e. you will have to create temporary states like semophores, which in the end will load the GC more than 208 bytes )

**Comment by uNetworkingAB** [comment]
151 ms to create 10 million slices does not sound like a problem. Say that you will be capped at 300k req/sec at best - that would put the total overhead of slicing at 0.4% if I did my math correctly. This is not a bottleneck, esp. not considering how sluggish JavaScript is in comparison.

**Comment by streamich** [comment]
The math is correct, assuming 300K req/sec, V8 will alloc/dealloc ~60 MB/sec unnecessarily and spend 0.45% or 4.6ms each second being blocked on creating slices.

**Comment by streamich** [comment]
Would you accept a PR if I implemented something like below?

```ts
interface WebSocket {
  sendSpecial(buffer: ArrayBuffer, offset: number, length: number): void;
}
```

**Comment by e3dio** [comment]
@streamich here is good benchmark I think

<details>
  <summary>Click to expand benchmark</summary>
  
  ### Heading
```javascript
const iterations = 1e7;
const ab = new ArrayBuffer(1024*4);
const buf = Buffer.from(ab);
const arr = new Uint8Array(ab);

console.time('Uint8Array-ab')
for(let i=0;i<iterations;i++){
	const pos=i%1024
	new Uint8Array(ab,pos,1)
}
console.timeEnd('Uint8Array-ab')

console.time('Uint8Array-arr')
for(let i=0;i<iterations;i++){
	const pos=i%1024
	new Uint8Array(arr,pos,1)
}
console.timeEnd('Uint8Array-arr')

console.time('Uint8Array-buf')
for(let i=0;i<iterations;i++){
	const pos=i%1024
	new Uint8Array(buf.buffer,pos+buf.byteOffset,1)
}
console.timeEnd('Uint8Array-buf')

console.time('subarray-arr')
for(let i=0;i<iterations;i++){
	const pos=i%1024
	arr.subarray(pos,pos+1)
}
console.timeEnd('subarray-arr')

console.time('subarray-buf')
for(let i=0;i<iterations;i++){
	const pos=i%1024
	buf.subarray(pos,pos+1)
}
console.timeEnd('subarray-buf')
```
</details>

Results:
> Uint8Array-ab: 247.ms (cheating: did not start from Buffer or Uint8Array)
Uint8Array-arr: 20s
Uint8Array-buf: 1.8s
subarray-arr: 1.3s
subarray-buf: 2.1s

So if you are starting from Buffer, `new Uint8Array()` is slightly faster than `subarray()` at 1.8s vs 2.1s. If you are starting from Uint8Array, subarray is 15x faster at 1.3s vs 20s. My encoding library uses Buffer so I would use `new Uint8Array()`

**Comment by uNetworkingAB** [comment]
If we're talking an overhead of 0.4% and ~200 bytes per slice I don't really see how this is even remotely an issue. It just sounds like a natural property of a scripting language. Scripting languages aren't zero overhead.

**Comment by e3dio** [comment]
The existing uWS.js API should be fine. Use subarray or new Uint8Array as described above to send uWS.js a zero copy slice/view of buffer

**Comment by uNetworkingAB** [comment]
There is a distinction between what is in scope and what is not. We can't add bypasses for custom use cases such as slice-and-send or extract-and-send or receive-to-json or anything like that. Then we would just add a bunch of hacks to workaround shortcomings of the language. Either accept the shortcomings of scripting or swap to the C++ or C library if you need guaranteed lowest overhead.

0.4% is not even a low hanging fruit

**Comment by streamich** [comment]
In my RPC requests and responses have schema, and encoder is compiled according to that schema, which writes straight to `ArrayBuffer`. The encoding is almost 10x faster than `Buffer.from(JSON.stringify(data))`, effectively this makes this last 0.5% bit of `.subarray()` the slowest part in the whole encoding process, which I wanted to eliminate.

**Comment by e3dio** [comment]
I think you have effectively optimized your process to the maximum, and you are safe to move on to something else ;) Also shameless plug check out my encoding library for smallest possible size, speed, and ease of use https://github.com/e3dio/packBytes Although I need to update the Buffer sizing mechanism, currently it calculates the exact buffer size needed instead of estimating and sending a view as described in this issue, I need to update that to improve speed

**Comment by streamich** [comment]
@e3dio thanks, yeah will take a look.

**Comment by uNetworkingAB** [comment]
> In my RPC requests and responses have schema, and encoder is JIT compiled according to that schema, which writes straight to big shared `Uint8Array`, so there are almost no allocations. Overall the encoding is almost 10x faster than `Buffer.from(JSON.stringify(data))`, effectively this makes this last 0.5% bit of `.subarray()` the slowest part in the whole encoding process, which I wanted to eliminate.

I totally salute your dedication but JavaScript is not the right place to be if you care about 0.4% slicing overhead.

**Comment by ronag** [comment]
There is also the overhead of GC of all the temporaries which might not be represented in the benchmarks...
